src/path-graph.js

```diff
@@ -376,44 +378,122 @@ export class Graph extends Component {
+      path = ['M', x1, y1, 'Q', qX, qY, x2, y2].join(' ');
+    }
+
+    // set edge path
     const edge = s[i];
-    d3.select(edge)
-      .attr('x1', d.source.x + sourceRadius * Math.cos(angle))
-      .attr('y1', d.source.y + sourceRadius * Math.sin(angle))
-      .attr('x2', d.target.x - targetRadius * Math.cos(angle))
-      .attr('y2', d.target.y - targetRadius * Math.sin(angle));
+    d3.select(edge).attr('d', path);
   }
 
   // position edge label in center of edge line and rotate
   positionEdgeLabel(d, i, s) {
-    const translateX = (d.source.x + d.target.x) / 2;
-    const translateY = (d.source.y + d.target.y) / 2;
-    let rotate =
-      (Math.atan2(d.target.y - d.source.y, d.target.x - d.source.x) * 180) /
-      Math.PI;
+    let x1 = d.source.x;
+    let y1 = d.source.y;
+    let x2 = d.target.x;
+    let y2 = d.target.y;
+
+    // get angle between source/target in radians
+    let angle = Math.atan2(y2 - y1, x2 - x1);
+
+    // get radius of source/target nodes
+    const sourceRadius = nodeRadius - 1;
+    let targetRadius = nodeRadius - 1;
+    // increase target node radius to bring tip of arrowhead out of circle
+    if (d.directed)
+      targetRadius += edgeArrowSize / 4;
+
+    // spread out contact points with circle over spread angle
+    const angleOffset = edgeSpreadAngle * d.coincidentOffset;
+
+    // bring start/end of curve to edge of circle
+    x1 += Math.cos(angle + angleOffset) * sourceRadius;
+    y1 += Math.sin(angle + angleOffset) * sourceRadius;
+    x2 -= Math.cos(angle - angleOffset) * targetRadius;
+    y2 -= Math.sin(angle - angleOffset) * targetRadius;
+
+    // get straight line distance between start/end of curve
+    const distance = Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
+
+    // get "sagitta" distance
+    const sag = Math.min(edgeSpreadDistance, distance) * d.coincidentOffset;
+
+    // get anchor point of text, point distance "sag" away from midpoint of line
+    const textX = (x2 + x1) / 2 - (sag * (y2 - y1)) / distance;
+    const textY = (y2 + y1) / 2 + (sag * (x2 - x1)) / distance;
+
+    // set vertical alignment of text relative to anchor point
+    let dy = -0.35 * edgeFontSize;
+    // always place text on "outside" side of curve
+    if (sag > 0 && d.source.x <= d.target.x)
+      dy = 0.85 * edgeFontSize;
+
+    // get angle of text in degrees
+    angle = (angle / (2 * Math.PI)) * 360;
+    // rotate text to always show upright
     if (d.source.x > d.target.x)
-      rotate += 180;
+      angle += 180;
 
+    // set edge text transform
     const edgeLabel = s[i];
     d3.select(edgeLabel)
       .attr('x', 0)
       .attr('y', 0)
+      .attr('dy', dy)
       .attr(
         'transform',
-        ' translate(' +
-          translateX +
-          ',' +
-          translateY +
-          ') ' +
-          ' rotate(' +
-          rotate +
-          ') '
+        ' translate(' + textX + ',' + textY + ') ' + ' rotate(' + angle + ') '
```

Comment:
This line generates a eslint warning:
```Unexpected string concatenation of literals  no-useless-concat```

It can be solved by changing the line into:
```
        ' translate(' + textX + ',' + textY + ')  rotate(' + angle + ') '
```
More details at: https://eslint.org/docs/rules/no-useless-concat

Reply:
Ah that's what that meant. I wanted to keep it split up like that in case I ever needed to change the order of the transformation (because the order matters). Maybe i'll just do something like `['translate(', textX, ',' ... ].join('');`

Reply:
Seems that CircleCI will pass if you fix this line.